test/unit-test/test_simulator.c

```diff
@@ -0,0 +1,368 @@
+    // specifically this line : task.as_mut().poll(context) in bb02_async.rs
+    usb_processing_process(usb_processing_hww());
+
+    // For getting the output packet, _send_next from hid_hww.c is called
+    // Flow : _send_next <- _sent_done <- hid_hww_setup <- _hww_endpoint_available
+    // <- usb_start <- _init_communication <- idle_workflow_blocking
+    //(These functions are all initialized in main function)
+}
+
+int main(void)
+{
+    char buffer[BUFFER_SIZE];
+    uint8_t input[BUFFER_SIZE];
+    int (*get_message)(char*, uint8_t*);
+    void (*send_message)(void);
+    bool memory_success, sd_success;
+    int temp_len;
+
+#ifdef SOCKET
+    int portno;
+    struct sockaddr_in serv_addr;
+    struct hostent* server;
+    portno = 15423;
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd < 0) {
+        perror("ERROR opening socket");
+        exit(1);
+    }
+    server = gethostbyname("host.docker.internal");
+    if (server == NULL) {
+        fprintf(stderr, "ERROR, no such host\n");
+        exit(0);
+    }
+    bzero((char*)&serv_addr, sizeof(serv_addr));
+    serv_addr.sin_family = AF_INET;
+    bcopy((char*)server->h_addr_list[0], (char*)&serv_addr.sin_addr.s_addr, server->h_length);
+    serv_addr.sin_port = htons(portno);
+    if (connect(sockfd, (struct sockaddr*)&serv_addr, sizeof(serv_addr)) < 0) {
+        perror("ERROR connecting");
+        exit(1);
+    }
+    get_message = get_usb_message_socket;
+    send_message = send_usb_message_socket;
+#elif PIPE
+    const char* fifo_read_path = "./tmp/fifo_sim_read";
+    const char* fifo_write_path = "./tmp/fifo_sim_write";
+    printf("opening fifo\n");
+    fd_read = open(fifo_read_path, O_RDONLY | O_NONBLOCK);
+    if (fd_read == -1) {
+        perror("open");
+        return 1;
+    }
+    printf("opened read fifo\n");
+    fd_write = open(fifo_write_path, O_WRONLY);
+    if (fd_write == -1) {
+        perror("open");
+        return 1;
+    }
+    printf("opened write fifo\n");
+    get_message = get_usb_message_pipe;
+    send_message = send_usb_message_pipe;
+#else
+    get_message = get_usb_message_stdin;
+    send_message = send_usb_message_stdout;
+#endif
+
+    /* When device starts, before firmware_main_loop, there are many initializations done,
+    they can be found in firmware.c. Apparently, not calling idle_workflow_blocking() causes
+    wrong address reads
+    */
+    // system_init();
+    usb_processing_init();
+    // For getting the output packet from the firmware
+    usb_processing_set_send(usb_processing_hww(), send_message);
+    printf("USB setup success\n");
+    hww_setup();
+    printf("HWW setup success\n");
+    // Memory initialization
+    sd_success = sd_format();
+    printf("Sd card setup %s\n", sd_success ? "success" : "failed");
+    mock_memory_factoryreset();
+    memory_interface_functions_t ifs = {
+        .random_32_bytes = _memory_setup_rand_mock,
```

Comment:
`= random_32_bytes_mcu` and `#include<random.h>` at the top, which uses C's `rand()` to get pseudorandom numbers, no need for the fixtures.